### ISSUE TITLE
.Net: Report function call failure in IFunctionFilter

### ISSUE THREAD

**Issue by Nurgo** [microsoft/semantic-kernel#4896]
`IFunctionFilter.OnFunctionInvoked()` allows us to be notified when the LLM has used a function successfully, but to my knowledge there is no way of knowing when the LLM has called a function that has failed (when an exception has been raised by the function, for example).

The solution would be that, in the event of an error, `IFunctionFilter.OnFunctionInvoked()` is also called, and provides the error message, or that another `OnFunctionFailed()` method is added to `IFunctionFilter`.

**Comment by dmytrostruk** [comment]
@Nurgo Thanks for reporting this issue! At the moment, as soon as you do `kernel.InvokeAsync(function)` and that `function` throws an error, you will be able to receive an error in place where you invoked a function, handle it with `try/catch` block and observe error message. Here is test example:
![image](https://github.com/microsoft/semantic-kernel/assets/13853051/dd0bf421-9494-4cbd-86f2-f27bc57d0510)

At the moment, `IFunctionFilter.OnFunctionInvoked()` is invoked only in case of successful function execution to observe the results, number of tokens used etc. Instead of handling exception in filters, it's possible to handle it with `try/catch`, which should be more native way for .NET. 

Please let me know if that makes sense. You could also provide specific scenario where you think exception details in filters will be useful, so we can see how to improve it. Thank you!

**Comment by matthewbolanos** [comment]
We may need to have an ability for a developer to have a more local "try/catch" for a function failure (for example, you may want to catch an error within a planner instead of putting a try/catch around the entire plan invocation). This will likely require new events.

To complexify things, according to @alliscode, the planners already have some logic to handle failures that would need to rationalize with.

@dmytrostruk, please sync with @gitri-ms in case this impacts the work she's doing with the new function calling filters.

**Comment by Nurgo** [comment]
@dmytrostruk Thank you for your example. The problem with this approach is that if the `KernelFuction` is not called explicitly, but by the LLM during a call to `IChatCompletionService.GetChatMessageContentAsync()`, it is not possible to know whether the exception was raised by a `KernelFunction`. It is therefore not possible to know the plugin and the function names in order to display a useful error message.

The `IFunctionFilter` will often be used to show the user when the chatbot is calling a function, so it would make sense to also be notified of errors, and not just successful function calls, as when a chatbot fails to use a plugin it's useful to inform the user.

**Comment by Nurgo** [comment]
When the exception is raised by a `KernelFunction` that has been called automatically by the LLM, during a call to `IChatCompletionService.GetChatMessageContentAsync()`, no exception is propagated, and the `ChatMessageContent` returned indicates that no function has been called, which is therefore incorrect.

As far as I know, there's no way of handling exceptions raised by `KernelFunctions` in this case.

`IFunctionFilter` seems to me a good place to handle errors in function calls, and should allow either to handle the exception locally and block its propagation, or to let it propagate to `GetChatMessageContentAsync()`.

What do you think?

**Comment by dmytrostruk** [comment]
@Nurgo Yes, I think that makes sense in case when function is called automatically by the LLM. Function filter should be a good place for it. Another option would be to add this feature to the new type of filters - tool filters. (PR: https://github.com/microsoft/semantic-kernel/pull/4922). These are the filters for functions that are called by LLM, we should check if it can cover this scenario.
